chronos/tunnels.nim

```diff
@@ -0,0 +1,228 @@
+  else:
+    discard
+
+  when defined(windows):
+    discard winlean.closeHandle(rtun.event)
+  else:
+    selectors.close(rtun.event)
+
+proc acquireLock(rtun: RawTunnel) {.inline.} =
+  ## Acquire lock in multi-threaded mode and do nothing in
+  ## single-threaded mode.
+  when hasThreadSupport:
+    acquire(rtun.lock)
+  else:
+    discard
+
+proc releaseLock(rtun: RawTunnel) {.inline.} =
+  ## Release lock in multi-threaded mode and do nothing in
+  ## single-threaded mode.
+  when hasThreadSupport:
+    release(rtun.lock)
+  else:
+    discard
+
+proc fireEvent(rtun: RawTunnel) {.inline.} =
+  ## Trigger OS event for Windows and SelectEvent for *nix.
+  when defined(windows):
+    discard setEvent(rtun.event)
+  else:
+    trigger(rtun.event)
+
+proc waitForEvent(rtun: RawTunnel): Future[bool] =
+  ## Wait asynchronously for OS event.
+  when defined(windows):
+    result = awaitForSingleObject(rtun.event, InfiniteDuration)
+  else:
+    result = awaitForSelectEvent(rtun.event, InfiniteDuration)
+
+proc newTunnel*[Msg](maxItems: int = -1): Tunnel[Msg] =
+  ## Create new Tunnel[Msg] with queue size ``maxItems``.
+  ##
+  ## If ``maxItems`` equal to ``-1`` (default value), then queue size is
+  ## unlimited.
+  let loop = getGlobalDispatcher()
+  result = cast[Tunnel[Msg]](allocShared(sizeof(RawTunnelImpl)))
+  result.initLocks()
+  result.mask = -1
+  result.maxItems = maxItems
+  result.count = 0
+  result.wr = 0
+  result.rd = 0
+  result.refCount = 0
+  result.data = cast[ptr UncheckedArray[byte]](0)
+
+proc open*[Msg](tun: Tunnel[Msg]) =
+  ## Open tunnel ``tun``.
+  let loop = getGlobalDispatcher()
+  tun.acquireLock()
+  inc(tun.refCount)
+  tun.releaseLock()
+
+proc close*[Msg](tun: Tunnel[Msg]) =
+  ## Close tunnel ``tun``.
+  tun.acquireLock()
+  if tun.refCount == 0:
+    if not(isNil(tun.data)):
+      deallocShared(cast[pointer](tun.data))
+    tun.deinitLocks()
+    deallocShared(cast[pointer](tun))
+  else:
+    dec(tun.refCount)
+    tun.releaseLock()
+
+proc `$`*[Msg](tun: Tunnel[Msg]): string =
+  ## Dump tunnel ``tun`` debugging information as string.
+  tun.acquireLock()
+  result = "Tunnel 0x" & toHex(cast[uint](tun)) & " ("
+  result.add("event = 0x" & toHex(cast[uint](tun.event)))
+  result.add(", rd = " & $tun.rd)
+  result.add(", wr = " & $tun.wr)
+  result.add(", count = " & $tun.count)
+  result.add(", mask = 0x" & toHex(tun.mask))
+  result.add(", data = 0x" & toHex(cast[uint](tun.data)))
+  result.add(", maxItems = " & $tun.maxItems)
+  result.add(", refCount = " & $tun.refCount)
+  result.add(")")
+  tun.releaseLock()
+
+proc rawSend(rtun: RawTunnel, pbytes: pointer, nbytes: int) =
+  var cap = rtun.mask + 1
+  if rtun.count >= cap:
+    if cap == 0: cap = 1
+    var n = cast[ptr UncheckedArray[byte]](allocShared0(cap * 2 * nbytes))
+    var z = 0
+    var i = rtun.rd
+    var c = rtun.count
+
+    while c > 0:
+      dec(c)
+      copyMem(addr(n[z * nbytes]), addr(rtun.data[i * nbytes]), nbytes)
```

Comment:
Can you add some comments explaining the logic here. Why are we copying the input data to multiple locations in the allocated buffer?

Reply:
@zah we maintain extendable queue in shared memory, here we copying data items from old queue to new queue (extended) because there no more room for new elements in old queue.